drivers/peci/peci_npcx.c

```diff
@@ -0,0 +1,279 @@
+};
+
+static int peci_npcx_check_bus_idle(struct peci_reg *reg)
+{
+	return IS_BIT_SET(reg->PECI_CTL_STS, NPCX_PECI_CTL_STS_START_BUSY) ? (-EBUSY) : 0;
+}
+
+static int peci_npcx_wait_completion(const struct device *dev)
+{
+	struct peci_npcx_data *const data = dev->data;
+	int ret;
+
+	ret = k_sem_take(&data->trans_sync_sem, PECI_TIMEOUT_MS);
+	if (ret != 0) {
+		LOG_ERR("%s: Timeout", __func__);
+		return -ETIMEDOUT;
+	}
+
+	if (data->trans_error != NPCX_PECI_NO_ERROR) {
+		return -EIO;
+	}
+	return 0;
+}
+static int peci_npcx_configure(const struct device *dev, uint32_t bitrate)
+{
+	const struct peci_npcx_config *const config = dev->config;
+	struct peci_npcx_data *const data = dev->data;
+	struct peci_reg *const reg = config->base;
+	uint8_t bit_rate_divider;
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	data->bitrate = bitrate;
+	/*
+	 * The unit of the bitrate is in Kbps, need to convert it to bps when
+	 * calculate the divider
+	 */
+	bit_rate_divider = ceiling_fraction(data->peci_src_clk_freq, bitrate * 1000 * 4) - 1;
+	/* Make sure the divider doesn't exceed the max valid value*/
+	bit_rate_divider = MIN(bit_rate_divider, NPCX_PECI_RATE_MAX_BIT_RATE_MASK);
+	/* Make sure the divider is not lower than the minimal valid value */
+	bit_rate_divider = MAX(bit_rate_divider, PECI_MAX_BIT_RATE_VALID_MIN);
+
+	if (bit_rate_divider < PECI_HIGH_SPEED_MIN_VAL) {
+		reg->PECI_RATE |= BIT(NPCX_PECI_RATE_EHSP);
+	} else {
+		reg->PECI_RATE &= ~BIT(NPCX_PECI_RATE_EHSP);
+	}
+	SET_FIELD(reg->PECI_RATE, NPCX_PECI_RATE_MAX_BIT_RATE, bit_rate_divider);
+
+	k_sem_give(&data->lock);
+
+	return 0;
+}
+
+static int peci_npcx_disable(const struct device *dev)
+{
+	struct peci_npcx_data *const data = dev->data;
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	irq_disable(DT_INST_IRQN(0));
+
+	k_sem_give(&data->lock);
+
+	return 0;
+}
+
+static int peci_npcx_enable(const struct device *dev)
+{
+	const struct peci_npcx_config *const config = dev->config;
+	struct peci_npcx_data *const data = dev->data;
+	struct peci_reg *const reg = config->base;
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	reg->PECI_CTL_STS = BIT(NPCX_PECI_CTL_STS_DONE) | BIT(NPCX_PECI_CTL_STS_CRC_ERR) |
+			    BIT(NPCX_PECI_CTL_STS_ABRT_ERR);
+	NVIC_ClearPendingIRQ(DT_INST_IRQN(0));
+	irq_enable(DT_INST_IRQN(0));
+
+	k_sem_give(&data->lock);
+
+	return 0;
+}
+
+static int peci_npcx_transfer(const struct device *dev, struct peci_msg *msg)
+{
+	const struct peci_npcx_config *const config = dev->config;
+	struct peci_npcx_data *const data = dev->data;
+	struct peci_reg *const reg = config->base;
+	struct peci_buf *peci_rx_buf = &msg->rx_buffer;
+	struct peci_buf *peci_tx_buf = &msg->tx_buffer;
+	enum peci_command_code cmd_code = msg->cmd_code;
+	int ret = 0;
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	ret = peci_npcx_check_bus_idle(reg);
+	if (ret != 0) {
+		goto out;
+	}
+
+	reg->PECI_ADDR = msg->addr;
+	reg->PECI_WR_LENGTH = peci_tx_buf->len;
+	reg->PECI_RD_LENGTH = peci_rx_buf->len;
```

Comment:
The PECI module has limits of 64 bytes for read and 65 bytes for write.  Please add checks for these limits.

Reply:
Done.